src/serac/numerics/functional/boundary_integral_kernels.hpp

```diff
@@ -164,56 +164,145 @@ void evaluation_kernel_impl(trial_element_type trial_elements, test_element, dou
-  auto          r   = reinterpret_cast<typename test_element::dof_type*>(outputs);
-  static constexpr TensorProductQuadratureRule<Q> rule{};
+  constexpr int dim                          = dimension_of(geom) + 1;
+  constexpr int nqp                          = num_quadrature_points(geom, Q);
+  using X_Type                               = const tensor<double, dim, nqp>;
+  using J_Type                               = const tensor<double, dim - 1, dim, nqp>;
+  auto                                     J = reinterpret_cast<J_Type*>(jacobians);
+  auto                                     x = reinterpret_cast<X_Type*>(positions);
+  auto                                     r = reinterpret_cast<typename test_element::dof_type*>(outputs);
+  constexpr TensorProductQuadratureRule<Q> rule{};
 
-  static constexpr int qpts_per_elem = num_quadrature_points(geom, Q);
+  constexpr int qpts_per_elem = num_quadrature_points(geom, Q);
 
   [[maybe_unused]] tuple u = {
       reinterpret_cast<const typename decltype(type<indices>(trial_elements))::dof_type*>(inputs[indices])...};
 
-  // for each element in the domain
-  for (uint32_t e = 0; e < num_elements; e++) {
-    // load the jacobians and positions for each quadrature point in this element
-    auto J_e = J[e];
-    auto x_e = x[e];
-
-    // batch-calculate values / derivatives of each trial space, at each quadrature point
-    [[maybe_unused]] tuple qf_inputs = {promote_each_to_dual_when<indices == differentiation_index>(
-        get<indices>(trial_elements).interpolate(get<indices>(u)[elements[e]], rule))...};
-
-    // (batch) evalute the q-function at each quadrature point
-    auto qf_outputs = batch_apply_qf(qf, t, x_e, J_e, get<indices>(qf_inputs)...);
-
-    // write out the q-function derivatives after applying the
-    // physical_to_parent transformation, so that those transformations
-    // won't need to be applied in the action_of_gradient and element_gradient kernels
-    if constexpr (differentiation_index != serac::NO_DIFFERENTIATION) {
-      for (int q = 0; q < leading_dimension(qf_outputs); q++) {
-        qf_derivatives[e * qpts_per_elem + uint32_t(q)] = get_gradient(qf_outputs[q]);
-      }
-    }
+  static constexpr trial_element_type empty_trial_element{};
+
+  using interpolate_out_type =
+      decltype(tuple{get<indices>(empty_trial_element).template interpolate_output_helper<Q>()...});
+
+  using qf_inputs_type = decltype(tuple{promote_each_to_dual_when<indices == differentiation_index>(
+      get<indices>(empty_trial_element).template interpolate_output_helper<Q>())...});
+
+#ifdef USE_CUDA
+  auto&           rm             = umpire::ResourceManager::getInstance();
+  auto            dest_allocator = rm.getAllocator("DEVICE");
+  qf_inputs_type* qf_inputs =
+      static_cast<qf_inputs_type*>(dest_allocator.allocate(sizeof(qf_inputs_type) * num_elements));
+  interpolate_out_type* interpolate_result =
```

Comment:
There's a lot of conditional compilation going on here that makes it harder to read these kernels-- and they were already pretty hard to read when I wrote them initially!

Can we investigate an alternate implementation that takes the device type as a template parameter, and use static polymorphism to specialize implementation details for each device instead of conditional compilation macros?

Reply:
I agree this is a good thing to do. However, in an effort to get this merged I say we table this for a follow on PR. Thoughts @white238 ?